### PR TITLE
Fix: Reader Search: Better "no results" state

### LINKED ISSUE #16603: Reader Search: Better "no results" state
Reader Search's "no results" state could use some love. Currently, at wide widths, we have POSTS and SITES columns even if one or both are empty:

Posts but no sites: https://wordpress.com/read/search?q=junotdiaz.com

screen shot 2017-07-26 at 2 14 52 pm

Sites but no posts: https://wordpress.com/read/search?q=powazek.co

screen shot 2017-07-26 at 2 15 22 pm

No results at all: https://wordpress.com/read/search?q=fishplops

screen shot 2017-07-26 at 1 46 12 pm

So what should happen? I suggest:

1. If there are Posts but no Sites, omit the Sites column and let Posts have the full width.

2. If there are Sites but no Posts, omit the Posts column and let Sites have the full width.

3. If there are no results at all, omit both headers (Posts and Sites) and center the "no results" message, changing the text to say "No posts or sites found for..."

Does this make sense? Is it doable?

cc @jancavan @gibrown @samouri

### PR DESCRIPTION
**Pull request by Saravanan90**
#### Changes proposed in this Pull Request

* When either of posts OR sites data is empty,
  * hides the respective column
  * makes the other column to occupy full width

* When both posts AND sites are empty,
  * displays the no results message
  * hides date/ relevance sorter

#### Testing instructions

* Results with posts but no sites

_Before fix (from the issue reported)_

Search results with posts only - before fix

_After fix_

Search results with posts only - after fix

* Results with sites but no posts

_Before fix (from the issue reported)_

Search results with sites only - before fix

_After fix_

Search results with sites only - after fix

* No search results

_Before fix (from the issue reported)_

No search results - before fix

_After fix_

No search results - after fix

Fixes https://github.com/Automattic/wp-calypso/issues/16603

### CODE REVIEW

**Comment by bluefuton** [comment]
Great! Thanks for tackling this @Saravanan90 👍

File: client/blocks/reader-combined-card/style.scss [hunk 1]
```diff
@@ -210,7 +210,6 @@
 	word-wrap: break-word;
 	position: relative;
 	margin-top: 4px;
-	white-space: nowrap;
```
Comment (bluefuton) [inline client/blocks/reader-combined-card/style.scss:213]: What's the rationale for removing this one?

File: client/reader/search-stream/index.jsx [hunk 2]
```diff
@@ -229,6 +236,9 @@ class SearchStream extends React.Component {
 								/>
 							</div>
 						) }
+						{ query && ! (showPostsColumn || showSitesColumn) &&
```
Comment (bluefuton) [inline client/reader/search-stream/index.jsx:239]: There are a few minor formatting issues that Prettier should fix for you (like the spacing inside the brackets here). Have you tried setting up wp-prettier (our Prettier fork) in your editor?

https://github.com/Automattic/wp-prettier

There's also a pre-commit hook that should do this formatting for you.

File: client/reader/search-stream/search-stream-header.jsx [hunk 3]
```diff
@@ -31,13 +33,17 @@ class SearchStreamHeader extends Component {
 	handleSitesSelected = () => this.props.onSelection( SEARCH_TYPES.SITES );
 
 	render() {
-		const { translate, wideDisplay, selected } = this.props;
+		const { translate, wideDisplay, selected, showPostsColumn, showSitesColumn } = this.props;
```
Comment (bluefuton) [inline client/reader/search-stream/search-stream-header.jsx:36]: Simple and clear - I like it 👍

File: client/reader/search-stream/index.jsx (hunk 2 above)
Comment (Saravanan90) [reply client/reader/search-stream/index.jsx:239]: @bluefuton I'll get this fixed.

File: client/blocks/reader-combined-card/style.scss (hunk 1 above)
Comment (Saravanan90) [reply client/blocks/reader-combined-card/style.scss:213]: @bluefuton When a result data has a long text, this css property prevented it from wrapping and made the content to overflow as evident in the below screenshot (with this css property applied)

With whitespace nowrap applied

Also please refer the below screenshot for the same result data after removing this

Without whitespace nowrap

File: client/reader/search-stream/index.jsx (hunk 2 above)
Comment (Saravanan90) [reply client/reader/search-stream/index.jsx]: This is fixed @bluefuton

**Comment by gibrown** [comment]
@Saravanan90 unfortunately I think hiding the "no results" messaging removes some important use cases:
1. Users searches for their own post and should see whether it does or does not show up
2. User searches for a specific site and sees whether they can follow it

I like the improvement where when there are no results for both then we collapse to a single "no results" view. I think we still do need to show some sort of no results for posts or sites when that is the case. IIRC in discussions on the original issue I think we were hoping to reduce the likelihood of the user hitting a deadend when they got no results. So we could show related searches, or related tags links beneath whatever told the user that there were no results. Unfortunately I am not sure we have any apis that would allow this to be implemented right now.

**Review by gibrown — CHANGES_REQUESTED** [review]
@Saravanan90 unfortunately I think hiding the "no results" messaging removes some important use cases:
1. Users searches for their own post and should see whether it does or does not show up
2. User searches for a specific site and sees whether they can follow it

I like the improvement where when there are no results for both then we collapse to a single "no results" view. I think we still do need to show some sort of no results for posts or sites when that is the case. IIRC in discussions on the original issue I think we were hoping to reduce the likelihood of the user hitting a deadend when they got no results. So we could show related searches, or related tags links beneath whatever told the user that there were no results. Unfortunately I am not sure we have any apis that would allow this to be implemented right now.

**Comment by Saravanan90** [comment]
@gibrown Thanks for the insight.
For the case when there are only posts without any sites for a particular search, an empty Sites column is displayed without any no results message conveyed to the user
_FYR_
Search results with posts only - before fix

Can you please clarify on the above?

**Note:** In case of search results with no posts, we do convey a no posts message to the user
_FYR_
Search results with posts only - before fix

**Comment by bluefuton** [comment]
Hi @Saravanan90! Based on Greg's feedback, let's make the following changes:

- If there are no results for _both_ columns, show the combined "no results" message. Otherwise, show both columns.
- Add a "no sites found" message to the Sites column. This will be shown when there are matching posts, but no matching sites.

**Comment by Saravanan90** [comment]
@bluefuton @gibrown Based on the last feedback, I have updated the fix and captured screenshots for all 3 use cases. Please verify them so that I can commit the changes made and submit for review

**No Posts**

No Posts

**No Sites**

No Sites

**No Results**

No Results

**Comment by bluefuton** [comment]
Thanks @Saravanan90! 

I suspect we might not want to use the full-size "no results" message in the sites column:

Screenshot

I'll ask for design review.

**Comment by Saravanan90** [comment]
Sure @bluefuton 👍 
I wasn't sure of which no results message to be displayed for sites column, hence reused the one which is currently displayed for posts column. Post design review, I'll make the suggested changes and submit for review.

**Comment by Saravanan90** [comment]
Hi @bluefuton !
May I know whether design review happened?

**Comment by bluefuton** [comment]
Hi @Saravanan90! Sorry for the delay. Attaching the "needs design review" label usually means a designer will drop by and take a look at the PR and they'll leave some comments. Let's give it a little while longer and I'll chase it up for you if we don't hear back in the next week or so 👍

**Comment by Saravanan90** [comment]
Thank you so much for clarifying @bluefuton !

**Comment by bluefuton** [comment]
@Saravanan90 I don't want to hold you up any further, so shall we just go with the text "No sites found.", centered at the top of the sites column?

**Comment by Saravanan90** [comment]
Sure @bluefuton I'll get the changes done and share the snap here

**Comment by sixhours** [comment]
👋 Hey hey @Saravanan90 ! Looks like this fell through the cracks. I still think it's a worthwhile improvement; if you want to give this a rebase/merge with trunk I'll take another look. Thank you very much for your contribution!

**Comment by bluefuton** [comment]
Thank you for your work on this @Saravanan90! ❤️

I think we'll opt to take a simpler approach and keep both columns visible at all times, but improve the "no results" messaging in the sites column. I'll work on a followup PR and be sure to give you credit for your part in this.

**Pull request closed without merging**